module/signature/aggregation.go

```diff
@@ -213,41 +230,40 @@ func (s *SignatureAggregatorSameMessage) Aggregate() ([]int, crypto.Signature, e
 
 // VerifyAggregate verifies an aggregated signature against the stored message and the stored
 // keys corresponding to the input signers.
-// Aggregating the keys of the signers internally is optimized to only look at the keys delta
-// compared to the latest execution of the function. The function is therefore not thread-safe.
+// The aggregated public key of input signers is returned.
+// The function is not thread-safe.
 // Possible returns:
-//   - (true, nil): aggregate signature is valid
-//   - (false, nil): aggregate signature is cryptographically invalid
-//   - (false, err) with error types:
+//   - (true, agg_key, nil): aggregate signature is valid
+//   - (false, agg_key, nil): aggregate signature is cryptographically invalid
+//   - (false, nil, err) with error types:
 //   - InsufficientSignaturesError if no signer indices are given (`signers` is empty)
 //   - InvalidSignerIdxError if some signer indices are out of bound
 //   - generic error in case of an unexpected runtime failure
-func (s *SignatureAggregatorSameMessage) VerifyAggregate(signers []int, sig crypto.Signature) (bool, error) {
-	sharesNum := len(signers)
-	keys := make([]crypto.PublicKey, 0, sharesNum)
-	if sharesNum == 0 {
-		return false, NewInsufficientSignaturesErrorf("cannot aggregate an empty list of signatures")
-	}
+func (s *SignatureAggregatorSameMessage) VerifyAggregate(signers []int, sig crypto.Signature) (bool, crypto.PublicKey, error) {
```

Comment:
do you have an opinion on moving the [check for identity public key](https://github.com/onflow/flow-go/blob/11d919c379660a1e866cf4b1eb3bc06c1a1f0cff/module/signature/aggregation.go#L219-L221) into this function? 

In essence, that would mean that the boolean return will only be false, if one of the input signatures is invalid. The case, where the aggregated key is identity is conceptually a setup error, where the `SignatureAggregatorSameMessage` has technically a dysfunctional configuration. From that angle it would make sense to handle these two scenarios differently on the API. 

I feel this would make the API a bit more expressive, and simpler 🤷 Alternatively, we could add a couple sentences to the method's goDoc explaining what a `false` return value implies and how to differentiate the two cases. (essentially describing [this](https://github.com/onflow/flow-go/blob/11d919c379660a1e866cf4b1eb3bc06c1a1f0cff/module/signature/aggregation.go#L217-L225))

Reply:
I gave the suggestion thoughts and I prefer the current "logic" of outputs, although I agree it simplifies the code a bit:

- IMHO, an error should result from a dysfunctional configuration or state. Verifying a signature against an identity key isn't a dysfunctional configuration. It is a normal setup that results in the function answering `false` to the question "is there enough guarantee that the private key corresponding to the input public key generated this input signature using this input message". If the signature has an invalid length or serialization, the answer to the question is `false` but no error is triggered. In the case of the identity public key, the answer is also `false` because the valid signature is the same for any message (i.e we can't be sure the private key has signed that specific input message), but there is no invalid setup of `SignatureAggregatorSameMessage`.  Therefore I don't fully agree to make `VerifyAggregate` return an error for this case. 
- Why would `Aggregate` return an error in this case? An aggregator intends to generate an aggregated signature that would eventually be valid when verified. If the aggregator fails to generate such valid signature, it makes sense for it to error.
- how would the caller of `VerifyAggregate` differentiate between an invalid signature because of invalid individual signature and identity public key? I think the function should provide extra data to differentiate those cases without errors, in this case I chose the aggregated key itself (but a simple boolean also works). I chose to output the key in case some future uses find it useful. 

I will add extra documentation for `VerifyAggregate` to clarify how to differentiate the subcases.

Reply:
> that would mean that the boolean return will only be false, if one of the input signatures is invalid

I also wanted to add that from the perspective of `VerifyAggregate`, there are no individual signatures. The function takes an input signature to verify, it doesn't consider any saved individual signature from `SignatureAggregatorSameMessage`, and ignores how that input signature was generated. The stateful object is only used to read the public keys we would like to test against.

Reply:
https://github.com/onflow/flow-go/pull/3441/commits/c192f795e1a6d214c7dfd1383c81c3a7727d4c71